instrumentation/rails/lib/opentelemetry/instrumentation/rails/span_subscriber.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Rails
+      # The SpanSubscriber is a special ActiveSupport::Notification subscription
+      # handler which turns notifications into generic spans, taking care to handle
+      # context appropriately.
+      class SpanSubscriber
+        def initialize(name:, tracer:)
+          @span_name = name.split('.')[0..1].reverse.join(' ').freeze
+          @tracer = tracer
+        end
+
+        def start(_name, _id, payload)
+          span = @tracer.start_span(@span_name, kind: :internal)
+          token = OpenTelemetry::Context.attach(
+            OpenTelemetry::Trace.context_with_span(span)
+          )
+
+          [span, token]
+        end
+
+        def finish(_name, _id, payload)
+          span = payload.delete(:__opentelemetry_span)
+          token = payload.delete(:__opentelemetry_ctx_token)
+          return unless span && token
+
+          attrs = payload.reject do |k, v|
+            %i[exception exception_object].include?(k) || v.nil?
+          end
+          span.add_attributes(attrs.transform_keys(&:to_s))
```

Comment:
Is there any risk here of the payload values (other than `exception` and `exception_object`, which you removed) having types other than String, Numeric or boolean?

Reply:
I don't have any strong feelings other than it shouldn't blow up 😄

Reply:
Actually, I do have one feeling I'd forgotten about...

One of the things we found very useful in our (much less impressive) ASN instrumentation was a "payload transform" capability. We allow subscribers to provide a denylist of payload keys to not include as attributes, and a transform lambda that takes the payload and returns a hash of attributes to add. We've used this to set the `peer.service` attribute, for example, and to remove things like the `:connection_id` and `:connection` object from `sql.active_record` spans.

Reply:
Makes sense. Check out 596239ead915f10be23fa4d2f20934fd8b5f6e45, 83e5fa06c62ec60afe11639ca66a62702ea3c1bc, and 3bedfda56112ac055bd04124da6b9dfd8db1a458 for the implementation!